### PR TITLE
SAAS-3837: Added AWS Config Delivery Failing Plugin

### PR DESCRIPTION
**Pull request by sadeed12345**

### CODE REVIEW

File: plugins/aws/configservice/configDeliveryFailing.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+            if (describeConfigurationRecorderStatus.err || !describeConfigurationRecorderStatus.data ||
+                !describeConfigurationRecorderStatus.data.length) {
+                helpers.addResult(results, 3,
+                    'Unable to query for Config Service status: ' + helpers.addError(describeConfigurationRecorderStatus), region);
+                return rcb();
+            }
+
+            for (let record of describeConfigurationRecorderStatus.data) {
+                if (record.lastStatus.toUpperCase()=='SUCCESS') {
```
Comment (m-nouman-suleman) [inline plugins/aws/configservice/configDeliveryFailing.js:33]:
```suggestion
                if (record.lastStatus.toUpperCase() === 'SUCCESS') {
```

File: plugins/aws/configservice/configDeliveryFailing.spec.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        "lastStatusChangeTime": "2022-01-19T12:19:56.700000+05:00"
+    }
+];
+
+const createCache = (status) => {
+    return {
+        configservice: {
+            describeConfigurationRecorderStatus: {
+                "us-east-1": {
+                    data: status                },
```
Comment (m-nouman-suleman) [inline plugins/aws/configservice/configDeliveryFailing.spec.js:26]:
```suggestion
                    data: status
                },
```

File: plugins/aws/configservice/configDeliveryFailing.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+                    helpers.addResult(results, 0,
+                        'The AWS Config service succeeded in delivering the last log file to the designated recipient.',
+                        region);
+                } else {
+                    helpers.addResult(results, 2,
+                        'The AWS Config service failed to deliver the last log file to the designated recipient.',
+                        region);
+                }
+            }
+            rcb();
```
Comment (m-nouman-suleman) [inline plugins/aws/configservice/configDeliveryFailing.js:43]:
```suggestion

            rcb();
```

File: plugins/aws/configservice/configDeliveryFailing.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+module.exports = {
+    title: 'Config Delivery Failing',
+    category: 'ConfigService',
+    domain: 'Management and Governance',
+    description: 'Ensure that the AWS Config log files are delivered without to the S3 bucket in order to store logging data for auditing purposes without any failures.',
+    more_info: 'Amazon Config keep record of the changes within the configuration of your AWS resources and it regularly stores this data to log files that are send to an S3 bucket specified by you.'+
+        'Sometimes you lose the ability to audit the configuration changes made within your AWS account this happen when AWS Config is not able to deliver log files to its recipient due to delivery errors or misconfigurations , the service is unable to send the recorded information to the designated bucket.',
+    recommended_action: 'Enable AWS Config log files to be delivered without any failures to designated S3 bucket, So that user could save logging data for auditing.',
+    link: 'https://docs.aws.amazon.com/config/latest/developerguide/stop-start-recorder.html',
```
Comment (m-nouman-suleman) [inline plugins/aws/configservice/configDeliveryFailing.js:12]: Please make sure this is the right link provided

File: plugins/aws/configservice/configDeliveryFailing.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    apis: ['ConfigService:describeConfigurationRecorderStatus'],
+
+    run: function(cache, settings, callback) {
+        var results = [];
+        var source = {};
+        var regions = helpers.regions(settings);
+        async.each(regions.configservice, function(region, rcb){
```
Comment (m-nouman-suleman) [inline plugins/aws/configservice/configDeliveryFailing.js:19]:
```suggestion
        
        async.each(regions.configservice, function(region, rcb){
```

File: plugins/aws/configservice/configDeliveryFailing.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+            if (!describeConfigurationRecorderStatus) return rcb();
+
+            if (describeConfigurationRecorderStatus.err || !describeConfigurationRecorderStatus.data ||
+                !describeConfigurationRecorderStatus.data.length) {
+                helpers.addResult(results, 3,
+                    'Unable to query for Config Service status: ' + helpers.addError(describeConfigurationRecorderStatus), region);
+                return rcb();
+            }
+
```
Comment (AkhtarAmir) [inline plugins/aws/configservice/configDeliveryFailing.js:32]: Add separate check for length

File: plugins/aws/configservice/configDeliveryFailing.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Config Delivery Failing',
+    category: 'ConfigService',
+    domain: 'Management and Governance',
+    description: 'Ensure that the AWS Config log files are delivered to the S3 bucket in order to store logging data for auditing purposes without any failures.',
+    more_info: 'Amazon Config keep record of the changes within the configuration of your AWS resources and it regularly stores this data to log files that are send to an S3 bucket specified by you.'+
+        'Sometimes you lose the ability to audit the configuration changes made within your AWS account this happen when AWS Config is not able to deliver log files to its recipient due to delivery errors or misconfigurations , the service is unable to send the recorded information to the designated bucket.',
```
Comment (AkhtarAmir) [inline plugins/aws/configservice/configDeliveryFailing.js:10]:
```suggestion
```

File: plugins/aws/configservice/configDeliveryFailing.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+                !describeConfigurationRecorderStatus.data.length) {
+                helpers.addResult(results, 3,
+                    'Unable to query for Config Service status: ' + helpers.addError(describeConfigurationRecorderStatus), region);
+                return rcb();
+            }
+
+            for (let record of describeConfigurationRecorderStatus.data) {
+                if (record.lastStatus.toUpperCase() === 'SUCCESS') {
+                    helpers.addResult(results, 0,
+                        'The AWS Config service succeeded in delivering the last log file to the designated recipient.',
```
Comment (AkhtarAmir) [inline plugins/aws/configservice/configDeliveryFailing.js:36]:
```suggestion
                        'AWS Config service is successfully delivering log files to the designated recipient',
```

File: plugins/aws/configservice/configDeliveryFailing.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+                        region);
+                } else {
+                    helpers.addResult(results, 2,
+                        'The AWS Config service failed to deliver the last log file to the designated recipient.',
```
Comment (AkhtarAmir) [inline plugins/aws/configservice/configDeliveryFailing.js:40]:
```suggestion
                        'AWS Config service is not delivering log files to the designated recipient successfully',
```

File: plugins/aws/configservice/configDeliveryFailing.js [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+            if (!describeConfigurationRecorderStatus) return rcb();
+
+            if (describeConfigurationRecorderStatus.err || !describeConfigurationRecorderStatus.data ||
+                !describeConfigurationRecorderStatus.data.length) {
+                helpers.addResult(results, 3,
+                    'Unable to query for Config Service status: ' + helpers.addError(describeConfigurationRecorderStatus), region);
+                return rcb();
+            }
+
```
Comment (AkhtarAmir) [inline plugins/aws/configservice/configDeliveryFailing.js:31]: Add check for length

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: plugins/aws/configservice/configDeliveryFailing.js [hunk 11]
```diff
@@ -0,0 +1,55 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Config Delivery Failing',
+    category: 'ConfigService',
+    domain: 'Management and Governance',
+    description: 'Ensure that the AWS Config log files are delivered to the S3 bucket in order to store logging data for auditing purposes without any failures.',
+    more_info: 'Amazon Config keep record of the changes within the configuration of your AWS resources and it regularly stores this data to log files that are send to an S3 bucket specified by you.',
+    recommended_action: 'Enable AWS Config log files to be delivered without any failures to designated S3 bucket, So that user could save logging data for auditing.',
```
Comment (AkhtarAmir) [inline plugins/aws/configservice/configDeliveryFailing.js:10]:
```suggestion
    recommended_action: 'Configure AWS Config log files to be delivered without any failures to designated S3 bucket.',
```

**Review by AkhtarAmir — APPROVED** [review]

**Pull request merged**